packages/@aws-cdk/aws-sns/lib/subscription.ts

```diff
@@ -215,3 +244,17 @@ export enum SubscriptionProtocol {
    */
   FIREHOSE = 'firehose'
 }
+
+/**
+ * The filter policy after it has had filters transformed by SubscriptionFilter
+ */
+interface SubscriptionFilterPolicyArrayWithMessageBody {
+  [attribute: string]: any[] | SubscriptionFilterPolicyArrayWithMessageBody;
+}
+
+/**
+ * The filter policy that allows for nested subscription filter policies. This is accomplished via recursive types.
+ */
+export interface SubscriptionFilterPolicyWithMessageBody {
+  [attribute: string]: SubscriptionFilter | SubscriptionFilterPolicyWithMessageBody;
```

Comment:
The main issue with the current way is that it breaks with the other languages when transpiled correct? The method   here with using a `FIlterOrPolicy` class would work but I it makes the user code very strange and unintuitive. I can do a little more research but we may have to go with this way right?

Reply:
Using this method, do you know how you would do DFS on this class with nested JSON objects? The keys are always defined as `json` making this part tricky.

Reply:
hey @brennanho can you give me an example of what you're trying to do (the equivalent in the existing code is fine if you have it working).

Reply:
hey @MrArnoldPalmer , basically I am trying to integrate @madeline-k 's `SubscriptionOptions` class with the DFS method I have.

```
export class FilterOrPolicy {
  public static filter(filter: SubscriptionFilter) {
    return new FilterOrPolicy(filter);
  }
  public static policy(policy: { [attribute: string]: FilterOrPolicy }) {
    return new FilterOrPolicy(policy);
  }

  private constructor(private readonly json: any) {}

  public toJSON() {
    return this.json;
  }
}
```

with 

```
  private buildFilterPolicyWithMessageBody(filterPolicy: any, depth = 1, totalCombinationValues = [1]): SubscriptionFilterPolicyArrayWithMessageBody {
    for (const [key, value] of Object.entries(filterPolicy)) {
      if (value instanceof SubscriptionFilter) {
        filterPolicy[key] = value.conditions;
        totalCombinationValues[0] *= value.conditions.length * depth;
      } else if (!(value instanceof Array)) {
        this.buildFilterPolicyWithMessageBody(value, depth + 1, totalCombinationValues);
      }
    }
    if (totalCombinationValues[0] > 150) {
      throw new Error(`The total combination of values (${totalCombinationValues}) must not exceed 150.`);
    }
    return filterPolicy;
  };
```

When I try change the types to correspond with the class (as suggested), all the key values when calling `Object.entries` are `json`. Also, calling `toJSON` on the object doesn't seem to work either since we need to convert every level of the object from class to basic json object. Do you know how you would do DFS on a class like this or perhaps is there another way to do this so we don't have to use a class or union types?

Reply:
Got it working!